File: py/0101_Symmetric_Tree.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def recursion(self, l, r):
        if (l and not r) or (not l and r):
            return False

        if not l and not r:
            return True

        if l.val != r.val:
            return False

        return self.recursion(l.left, r.right) and self.recursion(l.right, r.left)

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.recursion(root.left, root.right)


if __name__ == "__main__":
    solution = Solution()

    # Input: root = [1,2,2,3,4,4,3]
    # Output: true
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(3)
    print(solution.isSymmetric(root))

    # Input: root = [1,2,2,null,3,null,3]
    # Output: false
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)
    root.left.right = TreeNode(3)
    root.right.right = TreeNode(3)
    print(solution.isSymmetric(root))

    sys.exit(0)

File: cpp/0101_Symmetric_Tree.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool check(TreeNode* l, TreeNode* r) {
        if (!l && !r) return true;
        if ((l && !r) || (!l && r)) return false;
        if (l->val != r->val) return false;

        return check(l->left, r->right) && check(l->right, r->left);
    }

    bool isSymmetric(TreeNode* root) {
        return check(root->left, root->right);
    }
};

int main(int argc, char *argv[]) {

    return 0;
}
